src/octree/node_iterator.rs

```diff
@@ -0,0 +1,149 @@
+            intensity: None,
+        };
+
+        let edge_length = self.meta.bounding_cube.edge_length();
+        let min = self.meta.bounding_cube.min();
+        for _ in 0..self.meta.num_points {
+            // I tried pulling out this match by taking a function pointer to a 'decode_position'
+            // function. This replaces a branch per point vs a function call per point and turned
+            // out to be marginally slower.
+            match self.meta.position_encoding {
+                PositionEncoding::Float32 => {
+                    point.position.x = decode(
+                        self.xyz_reader.read_f32::<LittleEndian>().unwrap(),
+                        min.x,
+                        edge_length,
+                    );
+                    point.position.y = decode(
+                        self.xyz_reader.read_f32::<LittleEndian>().unwrap(),
+                        min.y,
+                        edge_length,
+                    );
+                    point.position.z = decode(
+                        self.xyz_reader.read_f32::<LittleEndian>().unwrap(),
+                        min.z,
+                        edge_length,
+                    );
+                }
+                PositionEncoding::Uint8 => {
+                    point.position.x =
+                        fixpoint_decode(self.xyz_reader.read_u8().unwrap(), min.x, edge_length);
+                    point.position.y =
+                        fixpoint_decode(self.xyz_reader.read_u8().unwrap(), min.y, edge_length);
+                    point.position.z =
+                        fixpoint_decode(self.xyz_reader.read_u8().unwrap(), min.z, edge_length);
+                }
+                PositionEncoding::Uint16 => {
+                    point.position.x = fixpoint_decode(
+                        self.xyz_reader.read_u16::<LittleEndian>().unwrap(),
+                        min.x,
+                        edge_length,
+                    );
+                    point.position.y = fixpoint_decode(
+                        self.xyz_reader.read_u16::<LittleEndian>().unwrap(),
+                        min.y,
+                        edge_length,
+                    );
+                    point.position.z = fixpoint_decode(
+                        self.xyz_reader.read_u16::<LittleEndian>().unwrap(),
+                        min.z,
+                        edge_length,
+                    );
+                }
+            }
+
+            point.color.red = self.rgb_reader.read_u8().unwrap();
+            point.color.green = self.rgb_reader.read_u8().unwrap();
+            point.color.blue = self.rgb_reader.read_u8().unwrap();
+            if let Some(ir) = self.intensity_reader.as_mut() {
+                point.intensity = Some(ir.read_f32::<LittleEndian>().unwrap());
+            }
+            f(&point);
+        }
+    }
+}
+
+fn fixpoint_decode<T>(value: T, min: f32, edge_length: f32) -> f32
```

Comment:
How about having `fixpoint_decode`, `decode`, `fixpoint_encode`, and `encode` in its own file, which is used in the iterator and writer?

Reply:
done

Reply:
Thanks Caterina.
How about renaming the file to `codec.rs`?

Reply:
ok! thank you for the suggestion